crates/dprint/src/commands/editor/mod.rs

```diff
@@ -117,105 +132,128 @@ impl<'a, TEnvironment: Environment> EditorService<'a, TEnvironment> {
   ) -> Self {
     let stdin = environment.stdin();
     let stdout = environment.stdout();
-    let reader_writer = StdIoReaderWriter::new(stdin, stdout);
+    let reader = MessageReader::new(stdin);
+    let writer = SingleThreadMessageWriter::for_stdout(MessageWriter::new(stdout));
+    let number_cores = environment.available_parallelism();
+    let concurrency_limiter = Arc::new(Semaphore::new(std::cmp::max(1, number_cores - 1)));
 
     Self {
-      messenger: StdIoMessenger::new(reader_writer),
+      reader,
       config: None,
       args,
       cache,
       environment,
       plugin_resolver,
       plugins_collection: plugin_pools,
+      context: Arc::new(EditorContext {
+        id_generator: Default::default(),
+        cancellation_tokens: Default::default(),
+        writer,
+      }),
+      concurrency_limiter,
     }
   }
 
   pub async fn run(&mut self) -> Result<()> {
     loop {
-      let message_kind = self.messenger.read_code()?;
-      match message_kind {
-        // shutdown
-        0 => return Ok(()),
-        // check path
-        1 => self.handle_check_path_message().await?,
-        // format
-        2 => self.handle_format_message().await?,
-        // unknown, exit
-        _ => bail!("Unknown message kind: {}", message_kind),
+      let message = EditorMessage::read(&mut self.reader)?;
+      // todo: temporary. Remove before merge or do a slimmed down version of it with log_verbose!(...)
+      eprintln!("Received message: {:?}", message);
```

Comment:
@ryan-rushton I added this message logging here and retested it with the vscode plugin and it's working.